### PR TITLE
[Keymap] Add ollyhayes personal keymap

### PR DESCRIPTION
**Pull request by ollyhayes**
## Description

Add my keymap for the crkbd keyboard

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/crkbd/keymaps/ollyhayes/README.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
+# ollyhayes's Keymap for [Corne Keyboard (crkbd)](https://github.com/foostan/crkbd)
```
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/README.md:1]: please rename this to be lower case.

File: keyboards/crkbd/keymaps/ollyhayes/config.h [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#define EE_HANDS
+
+// #define USE_SERIAL_PD2
```
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/config.h:21]:
```suggestion
```

File: keyboards/crkbd/keymaps/ollyhayes/config.h [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+#define TAPPING_FORCE_HOLD
+#define TAPPING_TERM 100
+
+#define NO_ACTION_ONESHOT
+#define NO_ACTION_MACRO
+#define NO_ACTION_FUNCTION
```
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/config.h:28]:
```suggestion
```

File: keyboards/crkbd/keymaps/ollyhayes/config.h [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+#ifdef RGBLIGHT_ENABLE
+    #undef RGBLED_NUM
+    #define RGBLIGHT_ANIMATIONS
```
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/config.h:32]: Could you explicitly define which animations are enabled? 
```suggestion
#define RGBLIGHT_EFFECT_BREATHING
#define RGBLIGHT_EFFECT_RAINBOW_MOOD
#define RGBLIGHT_EFFECT_RAINBOW_SWIRL
#define RGBLIGHT_EFFECT_SNAKE
#define RGBLIGHT_EFFECT_KNIGHT
#define RGBLIGHT_EFFECT_CHRISTMAS
#define RGBLIGHT_EFFECT_STATIC_GRADIENT
#define RGBLIGHT_EFFECT_RGB_TEST
#define RGBLIGHT_EFFECT_ALTERNATING
#define RGBLIGHT_EFFECT_TWINKLE
```

File: keyboards/crkbd/keymaps/ollyhayes/config.h [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    #define RGBLIGHT_VAL_STEP 17
+#endif
+
+#define OLED_FONT_H "keyboards/crkbd/lib/glcdfont.c"
+
+#ifdef RGB_MATRIX_ENABLE
+#   define RGB_MATRIX_KEYPRESSES // reacts to keypresses
+// #   define RGB_MATRIX_KEYRELEASES // reacts to keyreleases (instead of keypresses)
+// #   define RGB_DISABLE_AFTER_TIMEOUT 0 // number of ticks to wait until disabling effects
+#   define RGB_DISABLE_WHEN_USB_SUSPENDED true // turn off effects when suspended
```
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/config.h:46]:
```suggestion
#   define RGB_DISABLE_WHEN_USB_SUSPENDED // turn off effects when suspended
```

File: keyboards/crkbd/keymaps/ollyhayes/keymap.c [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
+#include "ollyhayes.h"
+
+#ifdef OLED_ENABLE
+#    include "oled.c"
```
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/keymap.c:21]: Please don't include c files.  You should use rules.mk to add it (`SRC += oled.c`).  Header files should be included, only.

File: keyboards/crkbd/keymaps/ollyhayes/oled.c [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    oled_write_ln(presses_str, false);
+}
+
+void oled_render_rgb_mode(void) {
+    static char mode_str[6] = {};
+    snprintf(
+        mode_str,
+        sizeof(mode_str),
+        "%u",
+        rgb_matrix_get_mode());
```
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/oled.c:107]:
```suggestion
```

File: keyboards/crkbd/keymaps/ollyhayes/oled.c [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+    oled_write_P(PSTR("Animation mode: "), false);
+    oled_write_ln(mode_str, false);
```
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/oled.c:110]:
```suggestion
    oled_write_ln(get_u8_str(rgb_matrix_get_mode(), ' '), false);
```

File: keyboards/crkbd/keymaps/ollyhayes/oled.c [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+    if (microticks++ % 1000 == 0) {
+        ticks++;
+    }
+
+    snprintf(
+        ticks_str,
+        sizeof(ticks_str),
+        "%u",
+        ticks);
```
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/oled.c:83]:
```suggestion
```

File: keyboards/crkbd/keymaps/ollyhayes/oled.c [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+        case L_DUBPUNC|L_PUNC:
+            oled_write_ln_P(PSTR("Specials"), false);
+            break;
+    }
+}
+
+void oled_render_timer(void) {
+    static uint16_t microticks = 0;
+    static uint16_t ticks = 0;
+    static char ticks_str[6] = {};
```
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/oled.c:73]:
```suggestion
```

File: keyboards/crkbd/keymaps/ollyhayes/oled.c [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    snprintf(
+        ticks_str,
+        sizeof(ticks_str),
+        "%u",
+        ticks);
+
+    oled_write_P(PSTR("Ticks: "), false);
+    oled_write_ln(ticks_str, false);
```
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/oled.c:86]:
```suggestion
    oled_write_ln(get_u16_str(ticks, ' '), false);
```

File: keyboards/crkbd/keymaps/ollyhayes/oled.c [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+void oled_render_presses(void) {
+    static char presses_str[6] = {};
+    snprintf(
+        presses_str,
+        sizeof(presses_str),
+        "%u",
+        key_presses);
```
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/oled.c:95]:
```suggestion
```

File: keyboards/crkbd/keymaps/ollyhayes/oled.c [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+
+    oled_write_P(PSTR("Presses: "), false);
+    oled_write_ln(presses_str, false);
```
Comment (drashna) [inline keyboards/crkbd/keymaps/ollyhayes/oled.c:98]:
```suggestion
    oled_write_ln(get_u16_str(presses_str, ' '), false);
```

File: users/ollyhayes/ollyhayes.c [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "ollyhayes.h"
+
+#include <stdio.h>
```
Comment (drashna) [inline users/ollyhayes/ollyhayes.c:19]:
```suggestion
```

File: users/ollyhayes/ollyhayes.c [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+#   include "virtser.h"
+#endif
+
+static void sendUtf(const char *str) {
+    // sequence for linux
+    register_code(KC_LCTRL);
+    register_code(KC_LSFT);
+    tap_code(KC_U);
+    unregister_code(KC_LSFT);
+    unregister_code(KC_LCTRL);
```
Comment (drashna) [inline users/ollyhayes/ollyhayes.c:30]:
```suggestion
    tap_code16(C(S(KC_U)));
```

File: users/ollyhayes/ollyhayes.c [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+    send_string(str);
+    tap_code(KC_ENTER);
+}
+
+layer_state_t default_layer_state_set_kb(layer_state_t state) {
+    if (state == 1UL<<BASE) {
```
Comment (drashna) [inline users/ollyhayes/ollyhayes.c:36]:
```suggestion
    if (layer_state_cmp(state, BASE)) {
```

File: users/ollyhayes/ollyhayes.c [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+        rgb_matrix_mode(3);
+    } else if (state == 1UL<<GAMES) {
```
Comment (drashna) [inline users/ollyhayes/ollyhayes.c:38]:
```suggestion
    } else if (layer_state_cmp(state,GAMES)) {
```

File: users/ollyhayes/ollyhayes.c [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+                tap_code(KC_DOWN);
+                return false;
+            }
+            break;
+
+        case CLEFT:
+            if (record->event.pressed) {
+                register_code(KC_LCTRL);
+                tap_code(KC_LEFT);
+                unregister_code(KC_LCTRL);
```
Comment (drashna) [inline users/ollyhayes/ollyhayes.c:86]: `C(KC_LEFT)`

File: users/ollyhayes/ollyhayes.c [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+                return false;
+            }
+            break;
+
+        case CRIGHT:
+            if (record->event.pressed) {
+                register_code(KC_LCTRL);
+                tap_code(KC_RIGHT);
+                unregister_code(KC_LCTRL);
```
Comment (drashna) [inline users/ollyhayes/ollyhayes.c:95]: `C(KC_RIGHT)`

File: users/ollyhayes/ollyhayes.c [hunk 20]
```diff
@@ -0,0 +1,203 @@
...
+ */
+
+#include "ollyhayes.h"
+
+#include <stdio.h>
+#ifdef VIRTSER_ENABLE
+#   include "virtser.h"
+#endif
+
+static void sendUtf(const char *str) {
```
Comment (drashna) [inline users/ollyhayes/ollyhayes.c:24]: Why not just use unicode?  You're essentially duplicating the code here, anyways. 

https://docs.qmk.fm/#/feature_unicode

**Review by drashna — CHANGES_REQUESTED** [review]

File: users/ollyhayes/ollyhayes.c (hunk 20 above)
Comment (drashna) [reply users/ollyhayes/ollyhayes.c:23]: Specifically, you can use `send_unicode_string(str)` for this instead.  You'd want to enable unicode, and you could use `UNICODE_COMMON = yes` in your rules, and just use the core unicode framework.  You'd just need to set the mode.

Comment (ollyhayes) [reply users/ollyhayes/ollyhayes.c:23]: Thanks! I switched to using the unicode feature and added the commit to the review.

**Comment by ollyhayes** [comment]
@drashna, I think I've addressed all your comments, and also added a few other things, and I rebased against the latest upstream/master. I've left them as separate commits so you can see what I did, shall I squash them all together?

I just have one other question, I never managed to get my 12 underglow leds to solder properly so I just left them out and shorted the data line, and then I changed `g_led_config`  in my branch to match the new set of leds like this: https://github.com/ollyhayes/qmk_firmware/commit/525cfdec3af10bd7ea793d65dcde8cb5afdc55dd

Obviously I can't submit this commit as it is, unless there an easy way to redefine this variable only for my keymap? If not I will just leave this commit only in my local repository.

**Review by drashna — APPROVED** [review]

File: keyboards/crkbd/keymaps/ollyhayes/keymap.c [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+#include QMK_KEYBOARD_H
+#include "ollyhayes.h"
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+  [BASE] = LAYOUT_split_3x6_3( \
+// +------------+------+------+----------+----------+---------+--------+----------+----------+--------+---------+---------+
+     NUM_SWITCH , KC_Q , KC_W , KC_F     , KC_P     , KC_G    , KC_J   , KC_L     , KC_U     , KC_Y   , KC_SCLN , KC_BSPC , \
+     KC_ESC     , KC_A , KC_R , KC_S     , KC_T     , KC_D    , KC_H   , KC_N     , KC_E     , KC_I   , KC_O    , KC_ENT  , \
+     KC_LSFT    , KC_Z , KC_X , KC_C     , KC_V     , KC_B    , KC_K   , KC_M     , KC_COMM  , KC_DOT , KC_SLSH , KC_RSFT , \
+                                KC_LCTRL , MO(PUNC) , MO(ARR) , KC_SPC , MO(PUNC) , KC_RCTRL                                \
```
Comment (fauxpark) [inline keyboards/crkbd/keymaps/ollyhayes/keymap.c:26]:
```suggestion
  [BASE] = LAYOUT_split_3x6_3(
// +------------+------+------+----------+----------+---------+--------+----------+----------+--------+---------+---------+
     NUM_SWITCH , KC_Q , KC_W , KC_F     , KC_P     , KC_G    , KC_J   , KC_L     , KC_U     , KC_Y   , KC_SCLN , KC_BSPC ,
     KC_ESC     , KC_A , KC_R , KC_S     , KC_T     , KC_D    , KC_H   , KC_N     , KC_E     , KC_I   , KC_O    , KC_ENT  ,
     KC_LSFT    , KC_Z , KC_X , KC_C     , KC_V     , KC_B    , KC_K   , KC_M     , KC_COMM  , KC_DOT , KC_SLSH , KC_RSFT ,
                                KC_LCTL , MO(PUNC) , MO(ARR) , KC_SPC , MO(PUNC) , KC_RCTL
```

File: keyboards/crkbd/keymaps/ollyhayes/keymap.c [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+                             KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS                                \
+// +---------+------+------+---------+---------+---------+---------+---------+---------+--------+---------+---------+
+  ),
+
+  [GAMES] = LAYOUT_split_3x6_3( \
+//  +---------+------+------+---------+---------+----------+---------+---------+---------+--------+---------+---------+
+	  KC_TRNS , KC_Q , KC_W , KC_E    , KC_R    , KC_T     , KC_Y    , KC_U    , KC_I    , KC_O   , KC_P    , KC_TRNS , \
+	  KC_TRNS , KC_A , KC_S , KC_D    , KC_F    , KC_G     , KC_H    , KC_J    , KC_K    , KC_L   , KC_SCLN , KC_TRNS , \
+	  KC_TRNS , KC_Z , KC_X , KC_C    , KC_V    , KC_B     , KC_N    , KC_M    , KC_COMM , KC_DOT , KC_SLSH , KC_TRNS , \
+							  KC_TRNS , KC_LALT , KC_SPACE , MO(ARR) , KC_TRNS , KC_TRNS                                \
```
Comment (fauxpark) [inline keyboards/crkbd/keymaps/ollyhayes/keymap.c:44]:
```suggestion
  [GAMES] = LAYOUT_split_3x6_3(
//  +---------+------+------+---------+---------+----------+---------+---------+---------+--------+---------+---------+
	  KC_TRNS , KC_Q , KC_W , KC_E    , KC_R    , KC_T     , KC_Y    , KC_U    , KC_I    , KC_O   , KC_P    , KC_TRNS ,
	  KC_TRNS , KC_A , KC_S , KC_D    , KC_F    , KC_G     , KC_H    , KC_J    , KC_K    , KC_L   , KC_SCLN , KC_TRNS ,
	  KC_TRNS , KC_Z , KC_X , KC_C    , KC_V    , KC_B     , KC_N    , KC_M    , KC_COMM , KC_DOT , KC_SLSH , KC_TRNS ,
							  KC_TRNS , KC_LALT , KC_SPACE , MO(ARR) , KC_TRNS , KC_TRNS
```

File: keyboards/crkbd/keymaps/ollyhayes/keymap.c [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+									   KC_DOT , KC_TRNS , KC_TAB    , KC_TRNS   , KC_0 , KC_DOT                              \
+//  +---------+---------------+------+--------+---------+-----------+-----------+------+--------+------+---------+---------+
+  ),
+
+  [FUNCTIONS] = LAYOUT_split_3x6_3( \
+//  +---------------+---------+---------+---------+---------+---------+---------+---------+---------+-------+---------+---------+
+	  KC_TRNS       , KC_TRNS , KC_TRNS , KC_F8   , KC_F9   , KC_F12  , KC_F12  , KC_F7   , KC_F8   , KC_F9 , KC_TRNS , KC_TRNS , \
+	  TG(FUNCTIONS) , KC_TRNS , KC_F4   , KC_F5   , KC_F6   , KC_F11  , KC_F11  , KC_F4   , KC_F5   , KC_F6 , KC_TRNS , KC_TRNS , \
+	  KC_TRNS       , KC_TRNS , KC_F1   , KC_F2   , KC_F3   , KC_F10  , KC_F10  , KC_F1   , KC_F2   , KC_F3 , KC_TRNS , KC_TRNS , \
+										  KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS                               \
```
Comment (fauxpark) [inline keyboards/crkbd/keymaps/ollyhayes/keymap.c:62]:
```suggestion
  [FUNCTIONS] = LAYOUT_split_3x6_3(
//  +---------------+---------+---------+---------+---------+---------+---------+---------+---------+-------+---------+---------+
	  KC_TRNS       , KC_TRNS , KC_TRNS , KC_F8   , KC_F9   , KC_F12  , KC_F12  , KC_F7   , KC_F8   , KC_F9 , KC_TRNS , KC_TRNS ,
	  TG(FUNCTIONS) , KC_TRNS , KC_F4   , KC_F5   , KC_F6   , KC_F11  , KC_F11  , KC_F4   , KC_F5   , KC_F6 , KC_TRNS , KC_TRNS ,
	  KC_TRNS       , KC_TRNS , KC_F1   , KC_F2   , KC_F3   , KC_F10  , KC_F10  , KC_F1   , KC_F2   , KC_F3 , KC_TRNS , KC_TRNS ,
										  KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS , KC_TRNS
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by ollyhayes** [comment]
I've rebased onto the latest master and squashed it all into one commit, hope that's okay!

I think I've address all the comments, is there anything else I should do before this could be merged?

Thanks for all your help :slightly_smiling_face:

**Review by keyboard-magpie — APPROVED** [review]

**Review by fauxpark — APPROVED** [review]

**Pull request merged**